### PR TITLE
The Big Ember Update

### PR DESCRIPTION
**Pull request by vevix**
TODO
---------
- [x] Update dependencies
- [x] Switch over to Ember 1.11.x `{{each}}` and `{{with}}` syntax where we can. (Some of this is going to require rewrites)
- [x] Convert all our `ObjectController` objects to `Controller`

What it means for you
-----------------------------

Ember 1.10+ uses HTMLBars :100: :tada: :tada: -- This means you can use a new binding syntax for HTML.

##### Before
```html
<div {{bind-attr class=":default someProperty:true:false"}}>
  <img {{bind-attr src=someOtherProperty}}>
</div>
```

##### After
```html
<div class="default {{if someProperty 'true' 'false'}}">
  <img src="{{someOtherProperty}}">
</div>
```

--------------------------------

If you are writing new code and need to use `{{each}}`, please use the new syntax.

##### Before
```html
{{#each para in model}}
  <p>{{para.stuff}}</p>
{{/each}}
```

##### After
```html
{{#each model as |para|}}
  <p>{{para.stuff}}</p>
{{/each}}
```

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -197,4 +197,7 @@
   end
 
   mount API => '/'
+
+  # Ember Tests
+  mount EmberCLI::Engine => 'ember-tests' if Rails.env.development?
```
Comment (NuckChorris) [inline config/routes.rb]: Why is this ember-tests in development and not test mode?

File: frontend/app/app.js [hunk 2]
```diff
@@ -25,9 +25,7 @@ Ember.Route.reopen({
   }
 });
 
-// Set to false because it breaks polymorphic associations.
-// Fixed in this PR: https://github.com/emberjs/data/pull/2586
-Ember.MODEL_FACTORY_INJECTIONS = false;
+Ember.MODEL_FACTORY_INJECTIONS = true;
```
Comment (NuckChorris) [inline frontend/app/app.js]: Isn't this actually the default? Do we even need to be explicit?

File: config/routes.rb (hunk 1 above)
Comment (vevix) [inline config/routes.rb]: I like to see the tests run! `/ember-tests/frontend`

File: frontend/app/app.js (hunk 2 above)
Comment (vevix) [inline frontend/app/app.js]: mm I looked and it was defaulted to false. It's possible I looked in the wrong place.

File: frontend/app/controllers/anime.js [hunk 3]
```diff
@@ -9,7 +9,7 @@ export default Ember.ObjectController.extend({
     var coverImage = this.get('model.coverImage');
     if (!coverImage) { return false; }
     var coverImageTopOffset = this.get('model.coverImageTopOffset');
-    return "background-image: url('" + coverImage + "'); background-position: 50% -" + coverImageTopOffset + "px;";
+    return (`background-image: url('${coverImage}'); background-position: 50% -${coverImageTopOffset}px;`).htmlSafe();
```
Comment (NuckChorris) [inline frontend/app/controllers/anime.js]: Might be a good place to abuse the fact that backtick strings are multi line

File: frontend/app/controllers/user.js [hunk 4]
```diff
@@ -2,22 +2,23 @@ import Ember from 'ember';
...
     return this.get('model.id') === this.get('currentUser.id');
   }.property('model.id'),
 
-  forumProfile: function () {
+  forumProfile: function() {
     return "https://forums.hummingbird.me/users/" + this.get('model.username');
   }.property('model.username'),
 
   // Legacy URLs
-  feedURL: function () {
```
Comment (NuckChorris) [inline frontend/app/controllers/user.js]: I think I removed both of these routes from `routes.rb` so these should probably get nuked too

Comment (vevix) [inline frontend/app/controllers/user.js]: Yeah I was thinking of cleaning up as I went but didn't. Radio should be nuked too :+1:

File: Gemfile [hunk 5]
```diff
@@ -20,7 +20,7 @@ gem 'redis', require: ['redis', 'redis/connection/hiredis']
 gem 'redis-rails'
 gem 'connection_pool'
 
-gem 'ember-cli-rails', '0.1.11'
+gem 'ember-cli-rails', '0.3.1'
```
Comment (NuckChorris) [inline Gemfile]: :pray: Here's hoping this bump fixes some of my issues

File: frontend/app/controllers/modals/edit-anime.js [hunk 6]
```diff
@@ -3,7 +3,7 @@ import ModalMixin from '../../mixins/modals/controller';
 import VersionableMixin from '../../mixins/modals/versionable';
 /* global moment */
 
-export default Ember.ObjectController.extend(ModalMixin, VersionableMixin, {
+export default Ember.Controller.extend(ModalMixin, VersionableMixin, {
   startingDate: Ember.computed('startedAiring', function(k, v) {
```
Comment (NuckChorris) [inline frontend/app/controllers/modals/edit-anime.js]: Shouldn't this be `model.startedAiring`? Same with down on 11 and 17 and 21, where it should probably be `model.finishedAiring`?

File: frontend/app/controllers/reviews/show.js [hunk 7]
```diff
@@ -1,22 +1,26 @@
...
-  coverImageStyle: function () {
-    return "background-image: url(" + this.get('user.coverImageUrl') + ")";
-  }.property('user.coverUrl'),
-  writtenBySelf: function () {
-    return this.get('currentUser.isSignedIn') && this.get('currentUser.id') === this.get('user.id');
-  }.property('user.id'),
-  ratingFirstDigit: function () {
+export default Ember.Controller.extend({
+  coverImageStyle: function() {
+    return (`background-image: url("${this.get('model.user.coverImageUrl')}")`).htmlSafe();
```
Comment (NuckChorris) [inline frontend/app/controllers/reviews/show.js]: Inlining the whole `get` looks a tad cluttered.  Maybe pull it out into a variable?

File: frontend/app/controllers/reviews/show.js [hunk 8]
```diff
@@ -1,22 +1,26 @@
...
+  }.property('model.user.coverUrl'),
```
Comment (NuckChorris) [inline frontend/app/controllers/reviews/show.js]: What's the difference between `model.user.coverUrl` and `model.user.coverImageUrl`?

File: frontend/app/controllers/reviews/show.js [hunk 9]
```diff
@@ -1,22 +1,26 @@
...
+  ratingDecimalPart: function() {
     return "." + (Math.floor((this.get('model.rating') - this.get('ratingFirstDigit')) * 10));
   }.property('model.rating'),
-  breakdown: function () {
+
+  breakdown: function() {
     var self = this,
         breakdown = [];
-    ['model.ratingStory', 'model.ratingAnimation', 'model.ratingSound', 'model.ratingCharacters', 'model.ratingEnjoyment'].forEach(function (property) {
+    ['model.ratingStory', 'model.ratingAnimation', 'model.ratingSound', 'model.ratingCharacters', 'model.ratingEnjoyment'].forEach(function(property) {
```
Comment (NuckChorris) [inline frontend/app/controllers/reviews/show.js]: #fatarrowopportunity

File: frontend/app/controllers/user/index.js [hunk 10]
```diff
@@ -3,11 +3,11 @@ import Ember from 'ember';
 export default Ember.ArrayController.extend({
   needs: "user",
   user: Ember.computed.alias('controllers.user'),
-  waifu_slug: Ember.computed.any('user.waifuSlug'),
-  hasWaifu: Ember.computed.any('user.waifu'),
-  hasLocation: Ember.computed.any('user.location'),
-  hasWebsite: Ember.computed.any('user.website'),
-  hasAboutText: Ember.computed.any('user.about'),
+  waifu_slug: Ember.computed.any('user.model.waifuSlug'),
```
Comment (NuckChorris) [inline frontend/app/controllers/user/index.js]: Underscores?  In _my_ JS?

It's more likely than you think!

`FREE PC CHECK`

File: frontend/app/models/user.js [hunk 11]
```diff
@@ -30,7 +30,7 @@ export default DS.Model.extend(ModelTruncatedDetails, {
   }.property('avatarTemplate'),
 
   coverImageStyle: function() {
```
Comment (NuckChorris) [inline frontend/app/models/user.js]: Holy shit how many times is this function repeated I've read this same thing at least 6 times at this point

File: frontend/app/templates/_notification.hbs [hunk 12]
```diff
@@ -1,21 +1,21 @@
...
-        {{sourceUser.username}}
+        {{notification.sourceUser.username}}
       </span>
 
-      {{#if isProfileComment}}
+      {{#if notification.isProfileComment}}
         wrote a comment on your profile.
       {{/if}}
-      {{#if isCommentReply}}
+      {{#if notification.isCommentReply}}
```
Comment (NuckChorris) [inline frontend/app/templates/_notification.hbs]: Note to @cybrox — have fun rebasing #426

File: frontend/app/templates/apps/index.hbs [hunk 13]
```diff
@@ -1,28 +1,25 @@
...
-              <a {{bind-attr href=link}}>{{name}}</a>
+            {{#each app.links as |link|}}
+              <a {{bind-attr href=link.link}}>{{link.name}}</a>
             {{/each}}
           </div>
           <div class="app-screens block-grid-md-2 block-grid-sm-1">
-            {{#each screenshots}}
+            {{#each app.screenshots as |screenshot|}}
               <div class="block-grid-item">
-                <img src="{{unbound this}}">
```
Comment (NuckChorris) [inline frontend/app/templates/apps/index.hbs]: Maybe this is a dumb question but why is the src unbound?  Isn't `unbound` mostly a performance thing? :u

File: frontend/app/templates/dashboard.hbs [hunk 14]
```diff
@@ -167,7 +167,11 @@
         </span>
       </div>
 
-      {{each itemController="story" itemViewClass="story"}}
+      {{#each model itemController="story" as |story|}}
+        {{#with story as |story|}}
```
Comment (NuckChorris) [inline frontend/app/templates/dashboard.hbs]: Isn't this redundant?

File: frontend/app/templates/group.hbs [hunk 15]
```diff
@@ -86,7 +86,7 @@
                     Approval Pending...
                   </button>
                 {{else}}
-                  <button class="btn btn-default btn-lg follow-button edit" {{action "leaveGroup"}}  {{bind-attr class="showEditMenu:hide"}}>
+                  <button class="btn btn-default btn-lg follow-button edit {{if showEditMenu 'hide'}}" {{action "leaveGroup"}}>
```
Comment (NuckChorris) [inline frontend/app/templates/group.hbs]: \>mixing two quote styles in the same line

File: frontend/app/templates/kotodama.hbs [hunk 16]
```diff
@@ -35,18 +35,18 @@
             <tbody>
               <tr><td>Registrations</td><td>{{accountsNew}}</td></tr>
```
Comment (NuckChorris) [inline frontend/app/templates/kotodama.hbs]: This isn't on `model`?

File: frontend/app/templates/user/following.hbs [hunk 17]
```diff
@@ -1,21 +1,20 @@
 <div class="container">
-  {{#each}}
+  {{#each model as |follower|}}
```
Comment (NuckChorris) [inline frontend/app/templates/user/following.hbs]: technically they'd be a `followee` not a `follower`.  Or what would be the opposite of `follower`?  `leader`?

File: frontend/app/templates/group.hbs (hunk 15 above)
Comment (wopian) [inline frontend/app/templates/group.hbs]: :ship it:

File: frontend/app/controllers/modals/edit-anime.js (hunk 6 above)
Comment (vevix) [inline frontend/app/controllers/modals/edit-anime.js]: :+1: good catch

File: frontend/app/controllers/reviews/show.js (hunk 8 above)
Comment (vevix) [inline frontend/app/controllers/reviews/show.js]: Haha just inconsistencies, on Anime/Manga we use `coverUrl` but User has `coverImageUrl`

Comment (NuckChorris) [inline frontend/app/controllers/reviews/show.js]: Both of these are on `user` though :stuck_out_tongue:

File: frontend/app/templates/dashboard.hbs (hunk 14 above)
Comment (vevix) [inline frontend/app/templates/dashboard.hbs]: Only way it works, can't pass story as content to a view ????? no idea why.

File: frontend/app/templates/kotodama.hbs (hunk 16 above)
Comment (vevix) [inline frontend/app/templates/kotodama.hbs]: Controller

File: frontend/app/templates/user/following.hbs (hunk 17 above)
Comment (vevix) [inline frontend/app/templates/user/following.hbs]: inc replace as `user` :trollface:

File: frontend/app/controllers/reviews/show.js (hunk 9 above)
Comment (vevix) [inline frontend/app/controllers/reviews/show.js]: I'm planning an ES6 PR soon that'll fix this all up!

File: frontend/app/templates/apps/index.hbs (hunk 13 above)
Comment (vevix) [inline frontend/app/templates/apps/index.hbs]: Yeah `unbound` doesn't create a binding so it just outputs and doesn't care after that. I'm okay with that here.

**Comment by NuckChorris** [comment]
With your fixes :+1: looks good.  I'd say wait for sven to maybe give a second set of eyes and call it good

File: frontend/app/controllers/reviews/show.js (hunk 8 above)
Comment (vevix) [inline frontend/app/controllers/reviews/show.js]: It was wrong and wasn't even used :/ I've pushed another change removing a few of these, as well as cleanup.

File: frontend/app/templates/_notification.hbs (hunk 12 above)
Comment (cybrox) [inline frontend/app/templates/_notification.hbs]: hmm

File: frontend/app/controllers/user/index.js (hunk 10 above)
Comment (cybrox) [inline frontend/app/controllers/user/index.js]: Really though, why isn't it just named `hasWaifuSlug`?

**Comment by cybrox** [comment]
I've read the word `model` in the past hour more than in the rest of my life, I guess.
Really wouldn't have seen anything odd other than the `waifu_slug` thingy. :+1:

File: frontend/app/controllers/user/index.js (hunk 10 above)
Comment (vevix) [inline frontend/app/controllers/user/index.js]: Haha pchen's work from the coffeescript days.

**Comment by vevix** [comment]
:+1: Gonna have another look over this myself and make sure I didn't miss anything.

**Pull request merged**